server/cluster/unsafe_recovery_controller.go

```diff
@@ -125,12 +128,181 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 }
 
 func (u *unsafeRecoveryController) isPlanExecuted(report *pdpb.StoreReport) bool {
-	return true
+	targetRegions := make(map[uint64]*metapb.Region)
+	toBeRemovedRegions := make(map[uint64]bool)
+	storeId := report.StoreId
+	for _, peerPlan := range u.storeRecoveryPlans[storeId].PeerPlan {
+		sourceRegionId := peerPlan.RegionId
+		if len(peerPlan.Targets) == 0 {
+			toBeRemovedRegions[sourceRegionId] = true
+			continue
+		}
+		for _, targetRegion := range peerPlan.Targets {
+			targetRegions[targetRegion.Id] = targetRegion
+		}
+	}
+	numFinished := 0
+	for _, peerReport := range report.Reports {
+		region := peerReport.RegionState.Region
+		if _, ok := toBeRemovedRegions[region.Id]; ok {
+			return false
+		} else if target, ok := targetRegions[region.Id]; ok {
+			if bytes.Equal(target.StartKey, region.StartKey) && bytes.Equal(target.EndKey, region.EndKey) {
+				numFinished += 1
+			} else {
+				return false
+			}
+		}
+	}
+	return numFinished == len(targetRegions)
+
+}
+
+type regionItem struct {
+	region *metapb.Region
+}
+
+func (r regionItem) Less(other btree.Item) bool {
+	return bytes.Compare(r.region.StartKey, other.(regionItem).region.StartKey) < 0
+}
+
+func (u *unsafeRecoveryController) canElectLeader(region *metapb.Region) bool {
+	numFailedPeers := 0
+	for _, peer := range region.Peers {
+		if _, ok := u.failedStores[peer.StoreId]; ok {
+			numFailedPeers += 1
+		}
+	}
+	return numFailedPeers*2 < len(region.Peers)
 }
 
 func (u *unsafeRecoveryController) generateRecoveryPlan() {
 	u.Lock()
 	defer u.Unlock()
+	regionReports := make(map[uint64]*pdpb.PeerReport)
+	storeByRegion := make(map[uint64]uint64)
+	for storeId, storeReport := range u.storeReports {
+		for _, peerReport := range storeReport.Reports {
+			regionId := peerReport.RegionState.Region.Id
+			if existing, ok := regionReports[regionId]; ok && existing.RaftState.LastIndex > peerReport.RaftState.LastIndex {
```

Comment:
After a second thought, we can't decide the newest region report based on `last_index`, since, in the case of partition, the old leader may have a bigger `last_index` though these logs can't be committed.

So compare epoch first, then last index.